### ISSUE TITLE
trade.set_custom_data(key, value) not persisting data between callbacks; trying to create TP with 2 stages

### ISSUE THREAD

**Issue by PTRPlay** [freqtrade/freqtrade#11756]
## Describe your environment

  * Operating system: docker
  * Python Version: Python 3.12.3
  * CCXT version: _____ (`pip freeze | grep ccxt`)
  * Freqtrade Version: freqtrade 2025.4
  * Mode: Backtesting
  
Note: All issues other than enhancement requests will be closed without further comment if the above template is deleted or not filled out.

## Describe the problem:

I'm encountering an issue when using **trade.set_custom_data(key, value)** inside a strategy during backtesting. The data set using this method does not persist callbacks. It appears that trade objects are re-instantiated on each callback, which causes the custom data to be lost. 

### Steps to reproduce:

  1. Inside **custom_exit()** or another callback, set custom data like:
      trade.set_custom_data("first_tp_hit", True)

  2. On the next callback, try to retrieve it (**adjust_trade_position**):
      trade.get_custom_data("first_tp_hit", False)

  3. The result is always the default (False), indicating the value was not stored.
  
### Expexcted Result:

  In backtesting mode, **trade.set_custom_data()** should persist the data for the lifetime of the trade, so that 
  custom state data can be tracked during the strategy logic.

### Code sample
I'm trying to develop a two stage take profit when the first TP is set to 1x of the Stop Loss and it closes 50% position size. Second TP should be 2x of SL and close remaining position. 
Here's the code:

``` python
class MACD_200SMA_Strategy(IStrategy):
    # Strategy interface version
    INTERFACE_VERSION = 3

    stoploss = -0.05
    # Minimal ROI table - defines points to take profit
    minimal_roi = {"0": 100}  # Effectively disable ROI table, using custom TP
    # Trailing stoploss
    trailing_stop = False
    position_adjustment_enable = True

    # Timeframe for the strategy
    timeframe = "1h"  # 1-hour candles                            

    macd_fast = CategoricalParameter([6, 8, 10, 12, 14, 16], default=12, space="buy", optimize=True)
    macd_slow = CategoricalParameter([18, 20, 22, 24, 26, 28, 30], default=26, space="buy", optimize=True)
    macd_signal = CategoricalParameter([7, 8, 9, 10, 11], default=9, space="buy", optimize=True)

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Calculates the indicators for our strategy
        """
        # MACD
        macd = ta.MACD(dataframe, fastperiod=self.macd_fast.value, slowperiod=self.macd_slow.value, signalperiod=self.macd_signal.value)

        dataframe["macd"] = macd["macd"]
        dataframe["macdsignal"] = macd["macdsignal"]
        dataframe["macdhist"] = macd["macdhist"]

        # Calculate MACD crossovers
        dataframe["macd_cross_up"] = qtpylib.crossed_above(dataframe["macd"], dataframe["macdsignal"])
        dataframe["macd_cross_down"] = qtpylib.crossed_below(dataframe["macd"], dataframe["macdsignal"])

        # Calculate 200 SMA
        dataframe["sma_200"] = ta.SMA(dataframe, timeperiod=200)

        # Calculate ATR
        dataframe["atr"] = ta.ATR(dataframe, timeperiod=14)

        # Price above/below 200 SMA
        dataframe["price_above_sma_200"] = dataframe["close"] > dataframe["sma_200"]  # todo: add ATR diff filter
        dataframe["price_below_sma_200"] = dataframe["close"] < dataframe["sma_200"]

        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Define entry conditions
        """
        dataframe.loc[
            (
                # Price is above 200 SMA (uptrend)
                dataframe["price_above_sma_200"]
                &
                # MACD crosses above signal line (bullish momentum)
                dataframe["macd_cross_up"]
                # MACD signal line < 0
                & (dataframe["macdsignal"] < dataframe["macdhist"])
            ),
            "enter_long",
        ] = 1

        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Define exit conditions
        """
        dataframe.loc[
            (
                # Price falls below 200 SMA (trend change)
                dataframe["price_below_sma_200"]
            ),
            ["exit_long", "exit_tag"],
        ] = [1, "close < sma_200"]

        return dataframe

    def custom_stake_amount(self, pair: str, current_time, current_rate: float, proposed_stake: float, min_stake: float, max_stake: float, **kwargs) -> float:
        # Use default stake amount
        return proposed_stake

    def custom_exit(self, pair: str, trade, current_time, current_rate: float, current_profit: float, **kwargs):
        """
        Custom exit logic for two-stage take profit
        - First TP: 50% at 1x stop-loss distance
        - Second TP: 50% at 2x stop-loss distance
        """
        # Get trade entry price
        entry_price = trade.open_rate
        stoploss_distance = abs(self.stoploss) * entry_price  # Absolute stop-loss distance in price

        # First take-profit: 1x stop-loss distance
        first_tp = entry_price + stoploss_distance / 2
        # Second take-profit: 2x stop-loss distance
        second_tp = entry_price + (stoploss_distance)

        # Check if first TP is hit and hasn't been closed yet
        if current_rate >= first_tp and not trade.get_custom_data("first_tp_hit", default=False):
            trade.set_custom_data(key="first_tp_hit", value=True)
            logger.warning(f"first_tp_hit set to True; open_rate: {trade.open_rate}")
            return "first_tp"  # Signal to close 50% of position

        # Check if second TP is hit
        if current_rate >= second_tp:
            return "second_tp"  # Signal to close remaining position

        return None

    def adjust_trade_position(self, trade, current_time, current_rate: float, current_profit: float, min_stake: float, max_stake: float, **kwargs):
        """
        Adjust position size when first TP is hit
        """
        logger.warning(f"adjust_trade_position; open_rate: {trade.open_rate}; first_tp_hit: {trade.get_custom_data('first_tp_hit')}")
        # Check if first TP was just hit using custom_data
        if trade.get_custom_data("first_tp_hit") and trade.amount > 0:
            logger.warning("Adjusting position size to 50% after first TP hit")
            current_amount = trade.amount
            return current_amount * 0.5  # Return half the current amount to close
        return None
```

**Comment by froggleston** [comment]
`return "first_tp"  # Signal to close 50% of position`
This will close the whole trade position immediately. custom_exit cannot be used for partial exits.

**Comment by xmatthias** [comment]
What froggleston tried to say (and i didn't understand initially, so now i'm certain the data remains there as i did test it, so letme try to rephrase this) is - you're setting the value 2 lines before returning a value from `custom_exit()` - which causes an immediate, full exit.

So once the value is set and your strategy returns `first_tp` - the trade will be closed - so `adjust_trade_position()` will not be called for this trade again.
So it's not a case of "isn't stored" - but "you set it, but never check again".

I think one sentence in [the docs](https://www.freqtrade.io/en/stable/strategy-callbacks/#custom-exit-signal) about `custom_exit()` describes behavior pretty well:

> Allows to define custom exit signals, indicating that specified position should be sold. 

(i'll reword this to "should be closed" to make it direction agnostic).
it does however make it clear that the whole position is closed (you're never telling the bot "only sell half" - that's not how custom_exit works). 
The `"first_tp"` response is the exit signal - but what you write here is irrelevant for bot behavior - as it's always going to result in a full exit.

**Comment by PTRPlay** [comment]
Thank you, everyone!
